quickwit/quickwit-serve/src/delete_task_api/handler.rs

```diff
@@ -0,0 +1,196 @@
+// For commercial licensing, contact us at hello@quickwit.io.
+//
+// AGPL:
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+use std::convert::Infallible;
+use std::sync::Arc;
+
+use quickwit_actors::Mailbox;
+use quickwit_janitor::actors::{DeleteTaskService, NewDeleteTask};
+use quickwit_metastore::Metastore;
+use quickwit_proto::metastore_api::DeleteQuery;
+use serde::Deserialize;
+use warp::{Filter, Rejection};
+
+use crate::format::Format;
+use crate::{require, with_arg};
+
+/// This struct represents the delete query passed to
+/// the rest API.
+#[derive(Deserialize, Debug, Eq, PartialEq, Default)]
+#[serde(deny_unknown_fields)]
+pub struct DeleteQueryRequest {
+    /// Query text. The query language is that of tantivy.
+    pub query: String,
+    // Fields to search on
+    #[serde(default)]
+    pub search_fields: Vec<String>,
+    /// If set, restrict delete to documents with a `timestamp >= start_timestamp`.
+    pub start_timestamp: Option<i64>,
+    /// If set, restrict delete to documents with a `timestamp < end_timestamp``.
+    pub end_timestamp: Option<i64>,
+}
+
+/// Delete query API handlers.
+pub fn delete_task_api_handlers(
+    metastore: Arc<dyn Metastore>,
+    delete_task_service_mailbox_opt: Option<Mailbox<DeleteTaskService>>,
+) -> impl Filter<Extract = impl warp::Reply, Error = Rejection> + Clone {
+    get_delete_tasks_handler(metastore.clone(), delete_task_service_mailbox_opt.clone()).or(
+        post_delete_tasks_handler(metastore, delete_task_service_mailbox_opt),
+    )
+}
+
+pub fn get_delete_tasks_handler(
+    metastore: Arc<dyn Metastore>,
+    delete_task_service_mailbox: Option<Mailbox<DeleteTaskService>>,
+) -> impl Filter<Extract = impl warp::Reply, Error = Rejection> + Clone {
+    warp::path!(String / "delete-tasks")
+        .and(warp::get())
+        .and(with_arg(metastore))
+        .and(require(delete_task_service_mailbox))
+        .and_then(get_delete_tasks)
+}
+
+// Returns delete tasks in json format for a given `index_id`.
+// Note that `_delete_task_service_mailbox` is not used...
+// Explanation: we don't want to expose any delete tasks endpoints without a running
+// `DeleteTaskService`. This is ensured by requiring a `Mailbox<DeleteTaskService>` in
+// `get_delete_tasks_handler` and consequently we get the mailbox in `get_delete_tasks` signature.
```

Comment:
I need to rewrite this comment.

Reply:
I thought it was very clear and super helpful / smart.